File: app/src/main/java/net/yupol/transmissionremote/app/server/AddServerActivity.java
package net.yupol.transmissionremote.app.server;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import net.yupol.transmissionremote.app.R;

public class AddServerActivity extends AppCompatActivity {

    public static final String PARAM_CANCELABLE = "param_cancelable";
    public static final String EXTRA_SEVER = "extra_server";

    private ServerDetailsFragment serverDetailsFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.add_server_activity);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setElevation(0);
        }

        FragmentManager fm = getSupportFragmentManager();
        serverDetailsFragment = (ServerDetailsFragment) fm.findFragmentById(R.id.add_server_fragment_container);
        if (serverDetailsFragment == null) {
            FragmentTransaction ft = fm.beginTransaction();
            serverDetailsFragment = new ServerDetailsFragment();
            ft.add(R.id.add_server_fragment_container, serverDetailsFragment);
            ft.commit();
        }

        Button okButton = findViewById(R.id.ok_button);
        okButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Server server = serverDetailsFragment.getNewServer();
                if (server != null) {
                    setResult(RESULT_OK, new Intent().putExtra(EXTRA_SEVER, server));
                    finish();
                }
            }
        });

        Button cancelButton = findViewById(R.id.cancel_button);
        boolean isCancelable = getIntent().getBooleanExtra(PARAM_CANCELABLE, true);
        cancelButton.setVisibility(isCancelable ? View.VISIBLE : View.GONE);
        if (isCancelable) {
            cancelButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    cancel();
                }
            });
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                cancel();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void cancel() {
        setResult(RESULT_CANCELED);
        finish();
    }
}

File: app/src/main/java/net/yupol/transmissionremote/app/server/AddServerActivity.kt
package net.yupol.transmissionremote.app.server

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import butterknife.ButterKnife
import butterknife.OnClick
import net.yupol.transmissionremote.app.R
import net.yupol.transmissionremote.app.TransmissionRemote
import net.yupol.transmissionremote.domain.repository.ServerRepository
import javax.inject.Inject

class AddServerActivity : AppCompatActivity() {

    companion object {
        private const val FRAGMENT_TAG = "fragment_tag"

        fun intent(context: Context): Intent {
            return Intent(context, AddServerActivity::class.java)
        }
    }

    @Inject lateinit var repo: ServerRepository

    override fun onCreate(savedInstanceState: Bundle?) {
        TransmissionRemote.getInstance().appComponent().inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.add_server_activity)
        ButterKnife.bind(this)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        if (savedInstanceState == null) {
            val fragment = ServerDetailsFragment.create()
            supportFragmentManager.beginTransaction()
                    .add(R.id.fragment_container, fragment, FRAGMENT_TAG)
                    .commit()
        }
    }

    @OnClick(R.id.ok_button)
    fun onOkButtonClicked() {
        val server = findFragment().newServer
        if (server != null) {
            repo.addServer(server)
            repo.setActiveServer(server)
            finish()
        }
    }

    private fun findFragment(): ServerDetailsFragment {
        return supportFragmentManager.findFragmentByTag(FRAGMENT_TAG) as ServerDetailsFragment
    }
}
